### DISCUSSION TITLE
Add a minimal runtime library in V langugage, remove the c lib depencecies.

### DISCUSSION THREAD

**Discussion by lygstate** [vlang/v#11598] [Ideas]
Usage: bare-metal environment
 And getting vlib depends on runtime lib, so vlib can be used under bare-metal env

**Comment by dcurrie** [comment]
The C library is often highly optimized for the target platform. So, it would be disadvantageous to remove it if it is available. Rather than removing the dependencies, perhaps it would be better to segregate or identify them to better support switching in an alternative for bare metal programming.

**Comment by medvednikov** [comment]
Using libc is ok. It's the only stable system API on macOS, Windows, Solaris etc. It will never be replaced there, just like in Go.

For bare metal it makes sense to avoid libc.

**Comment by lygstate** [comment]
libc on Windows are messy， have  Unicode and ansi problems

**Comment by medvednikov** [comment]
On Windows, Windows API is used where needed.

**Comment by lygstate** [comment]
On linux, I would suggest syscall instead libc to remove glibc dependencies

**Comment by kitech** [comment]
IMO, current libc for most populate system is good enough, many complex projects based on this and works perfect.

**Comment by medvednikov** [comment]
This is only really needed for bare-metal, mentioned in the original post.

For everything else libc is fine. So this is not a priority, unless someone is interested in working on this.

**Comment by kitech** [comment]
Agree with low priority. For bare-metal case, V would have an abstract libc layer, in case one day implemention it in V, or implemention it thirdparty. But even it's an abstract layer, it's still little big.

**Comment by MaD70** [comment]
In the light of other recent discussions on cutting dependencies: couldn't the transpiler from C to V be a (future) option for people who want zero dependencies?

I'm not say that there couldn't be (big) problems in translating system level C programs to V (and that people should be asking for such feature without investigating problems by themselves) but any experience with that could contribute to strengthening the transpiler.

**Comment by medvednikov** [comment]
@MaD70 do you mean separating the C to V translator from the compiler? 

That's already the case.

It will be installed separately, just like cross compilation files already do.

**Comment by MaD70** [comment]
@medvednikov no, I mean the transpiler could be used to translate from C to V system level programs like libc, loaders and alike.

**P.S.:** I'm thinking about the time when there will be various native code generators.

**Comment by medvednikov** [comment]
Translating system libc is not possible.

Something like musl would work, but it's Linux only.

**Comment by medvednikov** [comment]
Again, it's not possible to avoid libc/WinAPI on Windows, Solaris, and macOS. Not sure about BSDs.

**Comment by MaD70** [comment]
@medvednikov of course native APIs are unavoidable in the end, libc calls native APIs.

I can't imagine why the transpiler couldn't translate calls to native APIs (perhaps the problem is embedded assembly also?) Of course, if the native APIs are unstable one have to recompile the adapted system level library (say, libc) each time they change (but the advantage of a robust transpiler is that it's all automated). But we can postpone this discussion, I don't want to waste your time on a peripheral issue.

**Comment by crthpl** [comment]
This is being added with -freestanding, should this be closed only when that is finished?
